server/src/copilot/utils/lock.ts

```diff
@@ -0,0 +1,31 @@
+export class Lock {
+    private isLocked: boolean;
+    private waiting: Array<() => void>;
+
+    constructor() {
+        this.isLocked = false;
+        this.waiting = [];
+    }
+
+    acquire(): Promise<void> {
+        return new Promise((resolve) => {
+            if (!this.isLocked) {
+                this.isLocked = true;
+                resolve();
+            } else {
+                this.waiting.push(resolve);
+            }
+        });
+    }
+
+    release(): void {
+        if (this.waiting.length > 0) {
+            const nextResolve = this.waiting.shift();
+            if (nextResolve) {
+                nextResolve();
+            }
+        } else {
+            this.isLocked = false;
```

Comment:
What happens if someone has a bug in their code and doesn't call release?

Reply:
Fixed, introduced releaseTimeout to free out a thread if stuck.